daemon/logger/factory.go

```diff
@@ -19,16 +19,23 @@ type logdriverFactory struct {
 }
 
 func (lf *logdriverFactory) register(name string, c Creator) error {
-	lf.m.Lock()
-	defer lf.m.Unlock()
-
-	if _, ok := lf.registry[name]; ok {
+	if lf.driverRegistered(name) {
 		return fmt.Errorf("logger: log driver named '%s' is already registered", name)
 	}
+
+	lf.m.Lock()
 	lf.registry[name] = c
+	lf.m.Unlock()
 	return nil
 }
 
+func (lf *logdriverFactory) driverRegistered(name string) bool {
+	lf.m.Lock()
+	defer lf.m.Unlock()
```

Comment:
There's no reason to use `defer` here (it actually can make the function ~20% slower).